AAP-4364 - Restructure Builder Guide to create new EE Guide

dcdacosta opened the pull request:
This bug addresses the work for https://issues.redhat.com/browse/AAP-4364.

Changes in this pull request include:

- Restructure the builder guide and rename to Creating and consuming execution environments with Red Hat Ansible Automation Platform
- Include existing Builder information and also pulls in additional information from the PAH Guide.
- Adds two new topics for About EE and Why use EE as well as conceptual information where previously missing.

Preview link (VPN required): http://file.rdu.redhat.com/~ddacosta/AAP4364/master.html

thaumos commented:
@shanemcd can we get some eyes on this one from the team to provide some feedback?

Akasurde commented on downstream/modules/builder/con-about-builder.adoc:
```diff
@@ -11,4 +11,13 @@ Base the file name and the ID on the module title. For example:
-{Builder} is a command line tool that automates the process of building {ExecEnvName} by using the metadata defined in various Ansible Collections, as well as by the user.
+{Builder} is a command line tool that automates the process of building {ExecEnvName} by using the metadata defined in various Ansible Collections or by the user.
+
+Before {Builder} was developed, {PlatformName} users could run into dependency issues and errors when  creating custom virtual environments or containers that included all of the required dependencies installed.
```
```suggestion
Before {Builder} was developed, {PlatformName} users could run into dependency issues and errors when creating custom virtual environments or containers that included all of the required dependencies installed.
```

dcdacosta replied:
Change complete.

Akasurde commented on downstream/modules/builder/con-about-builder.adoc:
```diff
@@ -11,4 +11,13 @@ Base the file name and the ID on the module title. For example:
+
+Before {Builder} was developed, {PlatformName} users could run into dependency issues and errors when  creating custom virtual environments or containers that included all of the required dependencies installed.
+
+Now, with {Builder}, you can easily create a customizable {ExecEnvName} definition file that specifies the content you want included in your {ExecEnvName} such as, collections, requirements, and system level packages. This allows you fulfill all of the necessary requirements and dependencies to get jobs running.
```
```suggestion
Now, with {Builder}, you can easily create a customizable {ExecEnvName} definition file that specifies the content you want to be included in your {ExecEnvName} such as collections, third-party Python requirements, and system-level packages. This allows you to fulfill all of the necessary requirements and dependencies to get jobs running.
```

dcdacosta replied:
Change complete.

Akasurde commented on downstream/modules/builder/con-about-ee.adoc:
```diff
@@ -0,0 +1,21 @@
+An {ExecEnvName} is expected to contain the following:
+
+* Ansible 2.9 or Ansible Core 2.11
+* Python 3.8
```
```suggestion
* Python 3.8+
```

dcdacosta replied:
Change complete.

Akasurde commented on downstream/modules/builder/con-definition-dependencies.adoc:
```diff
@@ -1,6 +1,7 @@
 [id="con-definition-dependencies"]
 
 = Dependencies
+To avoid issues with your {ExecEnvNameSing} image, make sure your definition file adheres to all Galaxy, Python and system dependency requirements.
```
```suggestion
To avoid issues with your {ExecEnvNameSing} image, make sure your definition file adheres to all Galaxy, Python, and system dependency requirements.
```

Akasurde replied:
Do we want to mention "Automation Hub" instead of Galaxy?

dcdacosta replied:
I'm not sure whether to mention Hub versus Galaxy? Other changes complete.

Akasurde commented on downstream/modules/builder/con-ee-precedence.adoc:
```diff
@@ -0,0 +1,18 @@
+. The `default_environment` defined on the project that the job uses.
+. The `default_environment` defined on the organization of the job.
+. The `default_environment` defined on the organization of the inventory the job uses.
+. The current `DEFAULT_EXECUTION_ENVIRONMENT` setting (configurable at ``/api/v2/settings/jobs/`)
```
```suggestion
. The current `DEFAULT_EXECUTION_ENVIRONMENT` setting (configurable at ``/api/v2/settings/jobs/``)
```

dcdacosta replied:
Changes complete.

Akasurde reviewed (changes requested):
these are initial reviews, I will go in deep once these are fixed.

Akasurde commented on downstream/modules/builder/con-about-ee.adoc:
```diff
@@ -0,0 +1,21 @@
+
+[role="_abstract"]
+
+{ExecEnvNameStart} are container images on which all automation in {PlatformName} is run. {ExecEnvName} create a common language for communicating automation dependencies, and provide a standard way to build and distribute the automation environment.
```
<img width="964" alt="Screenshot 2022-08-30 at 12 59 15" src="https://user-images.githubusercontent.com/633765/187376540-e11aabd7-007f-4cd5-8ef0-d6db88b7be99.png">

{ExecEnvName} should start with a capital letter.

dcdacosta replied:
Change complete.

Akasurde commented on downstream/modules/builder/con-about-ee.adoc:
```diff
@@ -0,0 +1,21 @@
+
+{ExecEnvNameStart} are container images on which all automation in {PlatformName} is run. {ExecEnvName} create a common language for communicating automation dependencies, and provide a standard way to build and distribute the automation environment.
+
+An {ExecEnvName} is expected to contain the following:
```
`{ExecEnvName}` is plural

dcdacosta replied:
Changes complete.

Akasurde commented on downstream/modules/builder/con-why-ee.adoc:
```diff
@@ -0,0 +1,13 @@
+
+With {ExecEnvName}, {PlatformName} has transitioned to a distributed architecture by separating the control plane from the execution plane. Keeping automation execution independent of the control plane results in faster development cycles and improves scalability, reliability, and portability across environments. {PlatformName} also includes access to Ansible content tools, making it easy to buil ...
+
+In addition to speed, portability and flexibility, {ExecEnvName} provide the following benefits:
```
```suggestion
In addition to speed, portability, and flexibility, {ExecEnvName} provides the following benefits:
```

dcdacosta replied:
{ExecEnvName} is plural so provide is correct in this case. Other change complete.

Akasurde commented on downstream/modules/builder/con-about-ee.adoc:
```diff
@@ -0,0 +1,21 @@
+* Ansible 2.9 or Ansible Core 2.11
+* Python 3.8
+* Content collections
+* Collection, Python, or system dependencies
```
```suggestion
* Collection, Python, or system dependencies
* Container Runtime like Podman or Docker
```

dcdacosta replied:
Changes complete.

Akasurde commented on downstream/assemblies/builder/assembly-publishing-exec-env.adoc:
```diff
@@ -2,5 +2,6 @@
 
-include::builder/proc-push-execenv-to-hub.adoc[leveloffset=+1]
-include::builder/proc-pull-protected-registry.adoc[leveloffset=+1]
+include::builder/con-gathering-system-dependencies.adoc[leveloffset=+1]
```
I thinking publishing ee will only contain information about push and pull, not `con-gathering-system-dependencies.adoc` and `con-pip-based-req.adoc`.

dcdacosta replied:
push and pull are currently covered in another section "Modifying existing execution environments provided by AAP". Should they be duplicated here?

dcdacosta replied:
This may be fixed by the Google doc draft changes requested by @nitzmahone.

Akasurde commented:
cc @shrews @nitzmahone @eqrx

Akasurde commented on downstream/modules/builder/con-container_file.adoc:
```diff
@@ -2,12 +2,12 @@
 . In the final {ExecEnvName} image, system dependencies listed in the definition file are installed, including all system dependencies declared by collections listed in the definition file.
 . In the final {ExecEnvName} image, the downloaded collections are copied and the previously fetched Python dependencies are installed.
-. `append` for additional_build_steps from the definition file are run.
+. `append` for additional_build_steps from the definition file are run. 
```
```suggestion
. `append` for additional_build_steps from the definition file are run.
```

dcdacosta replied:
Change complete.

dcdacosta commented on downstream/assemblies/builder/assembly-building-off-existing-ee.adoc:
```diff
@@ -1,6 +1,8 @@
-include::builder/con-gathering-system-dependencies.adoc[leveloffset=+1]
-include::builder/proc-customize-ee-image.adoc[leveloffset=+1]
+{PlatformName} {PlatformVers} comes with a set of prebuilt, supported execution environments available on the Red Hat container registry. You can use these images, as is, for use in your environment or you can modify them to customize automation to meet your requirements.
+
```
Change added.

nitzmahone commented on downstream/modules/builder/ref-build-args-base-image.adoc:
```diff
@@ -9,14 +9,13 @@ The `build_arg_defaults` section of the definition file is a dictionary whose ke
 | `ANSIBLE_GALAXY_CLI_COLLECTION_OPTS`
-| * Allows the user to pass the `–pre` flag to enable the installation of pre-releases collections
-* `-c` is the equivalent of setting `verify_ssl` to `false`
+| * Allows the user to pass arbitrary arguments to the ansible-galaxy CLI during the collection installation phase. For example, the –pre flag to enable the installation of pre-release collections, or -c  to disable verification of the server's SSL certificate.
```
(the other values aren't bulleted, probably drop this bullet?)

dcdacosta replied:
Fixed.

nitzmahone commented:
Final result is much better- thanks!

dcdacosta commented on downstream/assemblies/builder/assembly-intro-to-builder.adoc:
```diff
@@ -1,9 +1,16 @@
+
+include::builder/con-about-ee.adoc[leveloffset=+1]
+include::builder/con-why-ee.adoc[leveloffset=+2]
+include::builder/con-ee-precedence.adoc[leveloffset=+2]
```
Moved to appendix.

dcdacosta commented on downstream/modules/builder/proc-customize-ee-image.adoc:
```diff
@@ -1,6 +1,6 @@
 [id="proc-customize-ee-image"]
 
-= Customizing an existing execution environment image
+= Customizing an existing {ExecEnvName} image
```
Verified for version 2.2. New issues will be created for future releases to ensure these are updated accordingly.

dcdacosta commented on downstream/modules/builder/con-about-ee.adoc:
```diff
@@ -0,0 +1,21 @@
+An {ExecEnvName} is expected to contain the following:
+
+* Ansible 2.9 or Ansible Core 2.11
+* Python 3.8
```
Left versions for now since they are correct for v2.2. Will consider creating a support matrix at some point to cover those specifics.

dcdacosta commented on downstream/modules/builder/con-about-ee.adoc:
```diff
@@ -0,0 +1,21 @@
+
+[role="_abstract"]
+
+{ExecEnvNameStart} are container images on which all automation in {PlatformName} is run. {ExecEnvName} create a common language for communicating automation dependencies, and provide a standard way to build and distribute the automation environment.
```
Fixed

ariordan-redhat commented on downstream/modules/builder/proc-customize-ee-image.adoc:
```diff
@@ -41,14 +41,17 @@ collections:
-NOTE: Since this example uses the community version of `kubernetes.core` and not a certified collection from {HubName}, we do not need to create an `ansible.cfg` nor reference that in our definition file.
+[NOTE]
+====
+Since this example uses the community version of `kubernetes.core` and not a certified collection from {HubName}, we do not need to create an `ansible.cfg` nor reference that in our definition file.
```
```suggestion
Since this example uses the community version of `kubernetes.core` and not a certified collection from {HubName}, we do not need to create an `ansible.cfg` file or reference that in our definition file.
```

ariordan-redhat commented on downstream/modules/builder/ref-build-args-base-image.adoc:
```diff
@@ -9,14 +9,13 @@ The `build_arg_defaults` section of the definition file is a dictionary whose ke
 
 | `EE_BASE_IMAGE`
-| Specifies the parent image for the {ExecEnvNameSing}, enabling a new image to be built that is based off of an already-existing image
+| Specifies the parent image for the automation execution environment, enabling a new image to be built that is based off of an already-existing image. This is typically a supported execution environment base image like ee-minimal or ee-supported, but can also be an execution environment image you've created previously and want to customize further.
```
```suggestion
| Specifies the parent image for the automation execution environment, enabling a new image to be built that is based off of an already-existing image. This is typically a supported execution environment base image like ee-minimal or ee-supported, but it can also be an execution environment image that you've created previously and want to customize further.
```

dcdacosta replied:
Fixed.

ariordan-redhat reviewed (commented):
A couple of suggestions: feel free to ignore them.

dcdacosta commented:
> The content looks good to me, but I had difficulty building the doc from the copy of the PR that I pulled down locally. It looks as if the symlink to the hub assemblies didn't get added to the PR. There's also a cross reference that fails because it links to a module that isn't included in the doc.

Thanks @ariordan-redhat , I think I added the symlink correctly and removed the broken xref/link. The file builds cleanly in my local repo so it should be ok now. Thanks for the review on this. It's ready to merge, backport and published if you're satisfied with these fixes. Once in pantheon, the associated Jira tickets can be closed as complete.

ariordan-redhat commented on downstream/titles/builder/master.adoc:
```diff
@@ -17,4 +17,10 @@ include::aap-common/making-open-source-more-inclusive.adoc[leveloffset=+1]
+include::hub/assembly-pull-image.adoc[leveloffset=+1]
+
+[appendix]
+include::../../modules/builder/con-ee-precedence.adoc[leveloffset=+1]
```
```suggestion
include::builder/builder/con-ee-precedence.adoc[leveloffset=+1]
```

ariordan-redhat commented:
Added 2 commits:

- Modified path to module in appendix to use the symlink names
- Deleted imagesdir setting in master.adoc to fix bad link from a hub module.
